x/sync/manager.go

```diff
@@ -510,43 +541,113 @@ func (m *Manager) handleRangeProofResponse(
+	if err := rangeProof.Verify(ctx); err != nil {
 		return err
 	}
 
-	if err := verifyRangeProof(
-		ctx,
-		&rangeProof,
-		int(request.KeyLimit),
-		maybeBytesToMaybe(request.StartKey),
-		maybeBytesToMaybe(request.EndKey),
-		request.RootHash,
-		m.tokenSize,
-		m.config.Hasher,
-	); err != nil {
-		return err
+	// Replace all the key-value pairs in the DB from start to end with values from the response.
+	if err := rangeProof.Commit(ctx); err != nil {
+		m.setError(err)
+		return nil
 	}
 
-	largestHandledKey := work.end
-
-	// Replace all the key-value pairs in the DB from start to end with values from the response.
-	if err := m.config.DB.CommitRangeProof(ctx, work.start, work.end, &rangeProof); err != nil {
+	nextKey, err := rangeProof.FindNextKey(ctx)
+	if err != nil {
 		m.setError(err)
 		return nil
 	}
+	m.completeWorkItem(work, nextKey, targetRootID)
+	return nil
+}
+
+func (p *rangeProof) Verify(ctx context.Context) error {
+	return verifyRangeProof(
+		ctx,
+		p.merkleProof,
+		p.request.keyLimit,
+		p.request.startKey,
+		p.request.endKey,
+		p.request.rootHash,
+		p.tokenSize,
+		p.hasher,
+	)
+}
+
+func (p *rangeProof) Commit(ctx context.Context) error {
+	if len(p.merkleProof.KeyChanges) == 0 {
+		return nil
+	}
+	return p.db.CommitRangeProof(ctx, p.request.startKey, p.request.endKey, p.merkleProof)
+}
 
-	if len(rangeProof.KeyChanges) > 0 {
-		largestHandledKey = maybe.Some(rangeProof.KeyChanges[len(rangeProof.KeyChanges)-1].Key)
+func (p *rangeProof) FindNextKey(ctx context.Context) (maybe.Maybe[[]byte], error) {
+	largestHandledKey := p.request.endKey
+	if len(p.merkleProof.KeyChanges) > 0 {
+		largestHandledKey = maybe.Some(p.merkleProof.KeyChanges[len(p.merkleProof.KeyChanges)-1].Key)
 	}
 
-	m.completeWorkItem(ctx, work, largestHandledKey, targetRootID, rangeProof.EndProof)
-	return nil
+	// Find the next key to fetch.
+	// If this is empty, then we have no more keys to fetch.
+	if !largestHandledKey.IsNothing() {
+		nextKey, err := findNextKey(ctx, p.db, largestHandledKey.Value(), p.request.endKey, p.merkleProof.EndProof, p.tokenSize)
+		if err != nil {
+			return maybe.Nothing[[]byte](), err
+		}
+		largestHandledKey = nextKey
+	}
+
+	return largestHandledKey, nil
+}
+
+func (m *merkleDBProofParser) RangeProof(responseBytes, rootHash []byte, startKey, endKey maybe.Maybe[[]byte], keyLimit uint32) (Proof, error) {
+	var rangeProofProto pb.RangeProof
+	if err := proto.Unmarshal(responseBytes, &rangeProofProto); err != nil {
+		return nil, err
+	}
+
+	var merkleProof merkledb.RangeProof
+	if err := merkleProof.UnmarshalProto(&rangeProofProto); err != nil {
+		return nil, err
+	}
+
+	return &rangeProof{
+		merkleProof: &merkleProof,
+		request: &rangeProofContext{
+			rootHash: rootHash,
+			startKey: startKey,
+			endKey:   endKey,
+			keyLimit: int(keyLimit),
+		},
+		db:        m.db,
+		tokenSize: m.tokenSize,
+		hasher:    m.config.Hasher,
+	}, nil
+}
+
+type rangeProofContext struct {
```

Comment:
nit: Should we replace the `*Context` suffix with `*Request` since that's where this data came from? I generally find that context is an overloaded term and doesn't indicate much information (like *backend.... and *manager...).